### PR TITLE
Introduce -emit-api-descriptor-path option

### PR DESCRIPTION
**Pull request by tshortli**
This option and its argument, when specified on the driver command line, should be passed through to jobs that emit modules and have access to the whole module.

Resolves rdar://116538520

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please test

File: Tests/SwiftDriverTests/SwiftDriverTests.swift [hunk 1]
```diff
@@ -7325,6 +7325,35 @@ final class SwiftDriverTests: XCTestCase {
...
+      let apiDescriptorPath = path.appending(component: "api.json").nativePathString(escaped: true)
+      var driver = try Driver(args: ["swiftc", "-whole-module-optimization",
+                                     "foo.swift", "bar.swift", "baz.swift",
+                                     "-module-name", "Test", "-emit-module",
+                                     "-emit-api-descriptor-path", apiDescriptorPath])
+
+      let jobs = try driver.planBuild().removingAutolinkExtractJobs()
+      let compileJob = try XCTUnwrap(jobs.first(where: {$0.kind == .compile}))
+
+      let supplementaryOutputs = try XCTUnwrap(compileJob.commandLine.supplementaryOutputFilemap)
```
Comment (cachemeifyoucan) [inline Tests/SwiftDriverTests/SwiftDriverTests.swift:7353]: does this always use file map file?

**Review by cachemeifyoucan — APPROVED** [review]
LGTM

File: Tests/SwiftDriverTests/SwiftDriverTests.swift (hunk 1 above)
Comment (tshortli) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:7353]: Apparently so with enough input files. This is the logic that determines whether or not an output file map will be used:

```
if inputsGeneratingCodeCount * FileType.allCases.count > fileListThreshold {
```

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
@swift-ci please test Windows

**Review by artemcm — APPROVED** [review]

File: Tests/SwiftDriverTests/SwiftDriverTests.swift (hunk 1 above)
Comment (tshortli) [reply Tests/SwiftDriverTests/SwiftDriverTests.swift:7352]: Ah, yeah I could do that.

**Comment by tshortli** [comment]
In the latest change I've also updated all the existing tests to use the new utility for extracting the supplementary file map from the command line.

**Comment by tshortli** [comment]
@swift-ci test

**Comment by tshortli** [comment]
@swift-ci test Windows

**Comment by tshortli** [comment]
@swift-ci test Windows

**Pull request merged**